Value formats

alaric-rd opened the pull request:
Herein is work to give the UI a way to select formats for columns that need them.

* An extra step ("format picking"), which can be omitted when no selected fields have types that need formats.
* The type guesser is used to highlight formats that match the supplied data.
* The request handler that sets field mappings can also set formats, so it can be used to provide a single form letting the user set both - or be used in two separate steps, as is done in the reference templates.
* The "basic" prototype has been updated to incorporate the new step.

The playwright tests are failing with a timeout - I can't run them locally so I'm not sure what they're upset about; adding an extra step to processes they click through, resulting in some expected button never turning up, might well be the problem. Guidance welcome!

alaric-rd committed:
Initial implementation of format picking in the UI.

Still to do:

 - Skip the format selection step if it's not needed
 - Improve text on the format selection step
 - Use guessed types to preselect formats if we can

alaric-rd committed:
Skip the formats page if no columns are mapped to attributes whose types require formats

alaric-rd committed:
Using guessed formats to guide the user in picking formats.

Includes a few bug fixes to the guesser!

alaric-rd committed:
Show the mapped field and type in the format picker page, as per the mockup

alaric-rd committed:
Appease linter

rossjones commented on lib/importer/src/functions.js:
```diff
@@ -48,10 +49,88 @@ const importerErrorMappingData = (error, key) => {
+        if (columnGuesses.types && columnGuesses.types.has(columnTypeName)) {
+            const guessedFormats = columnGuesses.types.get(columnTypeName);
+            if(guessedFormats && columnType.formats) {
+                likelyFormats = new Set(Array.from(columnType.formats.keys()).filter((fmtName) => {
```
Wonder why there isn't a Set.from() ? 

The filter function can have a short form anonymous function, so instead of 

```
.filter((fmtName) => {return guessedFormats.includes(fmtName)});
```

you can do this if there's only one expression..

```
.filter((fmtName) => guessedFormats.includes(fmtName) );
```

Just saves a couple of { } and a return ...

rossjones commented on lib/importer/src/functions.js:
```diff
@@ -257,3 +338,9 @@ module.exports = {
     data_avg
 }
+
+/*
+Local Variables:
+js-indent-level: 4
+End:
+*/
```
Is this intentional?

rossjones reviewed (approved):
Couple of suggestions, but I think it's looking good.  Not sure whether the formats page should show all of the fields as it gets a bit cluttered and hard to see the actual thing I need to do on the page for sheets like the pensions sample.

alaric-rd committed:
Fix playwright test

alaric-rd committed:
Code review feedback

alaric-rd commented on lib/importer/src/dudk/types/guesser.js:
```diff
@@ -64,6 +76,13 @@ exports.ListPossibleTypes = (field) => {
     // Either way, we can't narrow the set of possible types, but it's a blank cell
     return [false, true];
+  case "d": // Date or time or both
+    if(isDateFormat(field.z)) {
```
It should do, but I've put in a belt-and-braces check anyways.

alaric-rd commented on lib/importer/src/functions.js:
```diff
@@ -48,10 +49,88 @@ const importerErrorMappingData = (error, key) => {
+    const typeGuesses = backend_lib.SessionGuessTypes(session.backendSid, dataRange);
+
+    return Object.entries(mappings).map((entry) => {
+        const [colIndex, columnField] = entry;
```
Turned out you can't!

alaric-rd committed:
Fix bugs in review feedback (turns out you can't deconstruct things quite as hoped)

alaric-rd committed:
Review feedback: Don't imply separators in user-visible date format names

alaric-rd committed:
Review feedback (part 1): Extend type/format guesses to return both "safe" and "maybe" options

"Safe" options are valid for EVERY value in the column.
"Maybe" options are valid for SOME values in the column (but not every).

The frontend needs update to reflect this distinction, only showing
safe/maybe options (rather than ALL options as currently).

alaric-rd committed:
Update UI to reflect new distinction between safe/maybe/unsafe formats and types

simonwo commented:
Is the PR done with now?
